src/Simulation/TargetDefinitions/Decompositions/Utils.qs

```diff
@@ -166,21 +54,18 @@ namespace Microsoft.Quantum.Intrinsic {
         }
     }
 
-    /// Given a multiply-controlled operation that requires k controls 
-    /// applies it using ceiling(k/2) controls and using floor(k/2) temporary qubits
-    internal operation ApplyWithLessControlsA<'T> (op : ((Qubit[],'T) => Unit is Adj), (controls : Qubit[], arg : 'T)) : Unit is Adj {
-        let numControls = Length(controls);
-        let numControlPairs = numControls / 2;
-        use temps = Qubit[numControlPairs] {
-            within {
-                for numPair in 0 .. numControlPairs - 1 { // constant depth
-                    PhaseCCX(controls[2*numPair], controls[2*numPair + 1], temps[numPair]);
-                }
-            }
-            apply {
-                let newControls = numControls % 2 == 0 ? temps | temps + [controls[numControls - 1]];
-                op(newControls, arg);
-            }
+    internal operation CollectControls(ctls : Qubit[], aux : Qubit[]) : Unit is Adj {
```

Comment:
Some comments here may help explain what these two utility operations are intended for.

Reply:
Writing out the comments made realize the double controlled case worked but was using qubits inefficiently. I have a tweak that I think will correctly optimize those cases.

Reply:
Ok, I've added comments and adjusted the logic to correctly use fewer auxiliary qubits for doubly-controlled unitaries.